### ISSUE TITLE
new rule - function decorated with `@contextlib.contextmanager` that does not wrap its `yield` statement in a `try`/`finally` block

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#15629]
`contextlib.contextmanager` is dangerous because its `__exit__` won't get run if an exception is raised, which is rarely what you want:

```py
from contextlib import contextmanager

@contextmanager
def foo():
    print('starting')
    yield
    print('ending')

with foo():
    raise Exception
```

it would be nice if ruff had a rule to enforce `try`/`finally` on these types of context managers:

```py
from contextlib import contextmanager

@contextmanager
def foo():
    print('starting')
    try:
        yield
    finally:
        print('ending')
```

**Comment by AlexWaygood** [comment]
we would need to make sure that we also supported context managers that delegate to other context managers. E.g. the `g()` function definition is fine here:

```pycon
>>> from contextlib import contextmanager
>>> @contextmanager
... def f():
...     try:
...         print('hi')
...         yield
...     finally:
...         print('bye')
...         
>>> @contextmanager
... def g():
...     print('hiiiii')
...     with f():
...         yield
...         
>>> with g():
...     raise Exception
...     
hiiiii
hi
bye
Traceback (most recent call last):
  File "<python-input-3>", line 2, in <module>
    raise Exception
Exception
```

**Comment by DetachHead** [comment]
isn't that still unsafe if you have anything after the `yield` statement in `g`?

```py
@contextmanager
def g():
    print("hiiiii")
    with f():
        yield
        print("byeee")  # doesn't run
    print("byeee2")  # neither does this
```

i guess the rule can just not report on context managers where `yield` is the last statement:

```py
@contextmanager
def g():
    print("hiiiii")
    with f():
        yield # no error
```

```py
@contextmanager
def f():
    print("hi")
    yield # no error
```
but the last example probably isn't very common

**Comment by AlexWaygood** [comment]
> isn't that still unsafe if you have anything after the `yield` statement in `g`?

yes, fair point

> i guess the rule can just not report on context managers where `yield` is the last statement:

sure, that seems like a reasonable way to handle this.

---

Overall this seems like a reasonable rule to me, since you're right that this is easy to get wrong, and it's never _incorrect_ to put the `yield` inside a `try`/`finally` guard. Some might find it a little strict/noisy, since in some cases it isn't actually necessary to do the post-`yield` cleanup in the event that an exception is terminating the current scope's execution. But I do agree that it is strictly more correct to make sure that you always use `try`/`finally` in a `@contextmanager`-decorated function or method.

**Comment by MichaReiser** [comment]
Should we wait until after rule categorization with adding this rule, considering that it is rather noisy/strict.

**Comment by mishamsk** [comment]
I'd take a stab at this one if/when you confirm it should be taken

**Comment by anishgirianish** [comment]
Hi! I'd love to work on this if that's okay. Thanks!

**Comment by anishgirianish** [comment]
Hi! I've submitted a PR for this: [#22844](https://github.com/astral-sh/ruff/pull/22844)                                         
                                                                                          
The implementation includes:                                                            
- Detection of `@contextmanager` and `@asynccontextmanager` decorated functions with unprotected `yield`                                                                     
- Support for the delegation pattern (yield as last statement in a `with` block is not  flagged, per @AlexWaygood's example)                                                    
- Yield as the last statement in the function body is also not flagged, as suggested by  @DetachHead                                                                             
- The rule is in preview mode to address the strictness concern raised by @MichaReiser  
                                                                                          
Happy to make any adjustments based on feedback. Would appreciate a review when you get a chance. Thanks!
